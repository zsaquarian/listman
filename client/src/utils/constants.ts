export const LIST_SCHEMA_VERSION_NO = '1.2';
export const IGNORED_KEYS = ['master', 'config', 'auth-state'];
