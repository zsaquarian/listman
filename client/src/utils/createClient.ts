import { dedupExchange, fetchExchange, initClient } from '@urql/svelte';
import { cacheExchange } from '@urql/exchange-graphcache';
import { MeDocument, RefreshDocument } from '../graphql';
import type { CreateUserMutation, GoogleSignInMutation, LoginMutation } from '../graphql';
import { authExchange } from '@urql/exchange-auth';
import { authStore } from '@store/auth';

export const createClient = (): void => {
  initClient({
    url: import.meta.env.VITE_GRAPHQL_IP,
    fetchOptions: {
      credentials: 'include',
    },
    exchanges: [
      dedupExchange,
      cacheExchange({
        updates: {
          Mutation: {
            login: (result: LoginMutation, _args, cache, _info) => {
              const user = result.login.user;

              cache.updateQuery({ query: MeDocument }, (data) => {
                if (result.login.error) {
                  return data;
                } else {
                  authStore.update((val) => ({ ...val, isLoggedIn: true }));
                  return {
                    me: {
                      id: user.id,
                      uuid: user.uuid,
                      email: user.email,
                      username: user.username,
                    },
                  };
                }
              });
            },
            googleSignIn: (result: GoogleSignInMutation, _args, cache, _info) => {
              const user = result.googleSignIn.user;

              authStore.update((val) => ({ ...val, isLoggedIn: true }));
              cache.updateQuery({ query: MeDocument }, (data) => {
                if (result.googleSignIn.error) {
                  return data;
                } else {
                  return {
                    me: {
                      id: user.id,
                      uuid: user.uuid,
                      email: user.email,
                      username: user.username,
                    },
                  };
                }
              });
            },
            createUser: (result: CreateUserMutation, _args, cache, _info) => {
              const user = result.createUser.user;

              authStore.update((val) => ({ ...val, isLoggedIn: true }));
              cache.updateQuery({ query: MeDocument }, (data) => {
                if (result.createUser.error) {
                  return data;
                } else {
                  return {
                    me: {
                      id: user.id,
                      uuid: user.uuid,
                      email: user.email,
                      username: user.username,
                    },
                  };
                }
              });
            },
            logOut: (_result, _args, cache, _info) => {
              authStore.update((val) => ({ ...val, isLoggedIn: false }));

              cache.updateQuery({ query: MeDocument }, (_data) => {
                return {
                  me: null,
                };
              });
            },
          },
        },
      }),
      authExchange({
        addAuthToOperation: ({ operation }) => {
          return operation;
        },
        didAuthError: ({ error }) => {
          return error.graphQLErrors.some((val) => val.message.includes('Not authorized'));
        },
        getAuth: async ({ authState, mutate }) => {
          if (!authState) {
            return {
              loggedIn: false,
            };
          }

          const result = await mutate(RefreshDocument);

          if (result.data?.refresh?.token) {
            return {
              loggedIn: true,
            };
          }

          return {
            loggedIn: false,
          };
        },
      }),
      fetchExchange,
    ],
  });
};
