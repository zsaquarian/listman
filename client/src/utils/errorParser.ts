import type { CombinedError } from '@urql/svelte';
import type { GraphQLError } from 'graphql';

export interface ParsedError {
  type: 'MissingParam' | 'InvalidParam' | 'Unknown';
  field?: string;
}

const fieldMap = {
  usernameOrEmail: 'Username or e mail',
  password: 'Password',
  username: 'Username',
  email: 'E mail',
};

const missingParamRegex = /Variable "\$(.*)" of/;
const invalidParamRegex = /(.*) does not comply/;

export const parseError = (err: GraphQLError): ParsedError => {
  // Missing param
  if (err.message.includes('required type')) {
    const missingParamReg = err.message.match(missingParamRegex);

    return {
      type: 'MissingParam',
      field: missingParamReg[1],
    };
  }

  // Invalid param
  if (err.message.includes('comply with requirements')) {
    const invalidParamReg = err.message.match(invalidParamRegex);

    return {
      type: 'InvalidParam',
      field: invalidParamReg[1],
    };
  }

  return {
    type: 'Unknown',
  };
};

export const parseCombinedError = (err: CombinedError): ParsedError[] => {
  const allErrors = [] as ParsedError[];

  err.graphQLErrors.forEach((val) => {
    allErrors.push(parseError(val));
  });

  return allErrors;
};

export const getReadableError = (err: ParsedError): string => {
  if (err.type === 'MissingParam') return 'Please enter your ' + fieldMap[err.field];
  if (err.type === 'InvalidParam') return 'Field "' + fieldMap[err.field] + '" is invalid';

  return 'An unknown error occured. Please try later.';
};
