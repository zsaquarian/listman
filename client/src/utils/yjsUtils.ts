import { syncedStore, getYjsValue } from '@syncedstore/core';
import type { GenericList } from './listUtils';
import { svelteSyncedStore } from '@syncedstore/svelte';
import { WebsocketProvider } from 'y-websocket';
import type { Readable } from 'svelte/store';
import type { MappedTypeDescription } from '@syncedstore/core/types/doc';

export const getStore = (key: string): Readable<MappedTypeDescription<{ todos: GenericList }>> => {
  const listStore = syncedStore({ todos: {} as GenericList });
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new WebsocketProvider(import.meta.env.VITE_YJS_IP, key, getYjsValue(listStore) as any); // sync via websocket
  const store = svelteSyncedStore(listStore);

  return store;
};
