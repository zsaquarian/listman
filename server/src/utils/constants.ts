import dotenv from 'dotenv';

dotenv.config();
export const PORT = parseInt(process.env.PORT || '3000');
export const GOOGLE_CLIENT_ID_WEB = process.env.GOOGLE_CLIENT_ID_WEB;
export const GOOGLE_CLIENT_ID_ANDROID = process.env.GOOGLE_CLIENT_ID_ANDROID;
export const JWT_SECRET = process.env.JWT_SECRET;
export const REFRESH_SECRET = process.env.REFRESH_SECRET;
export const CORS_ORIGIN = process.env.CORS_ORIGIN;

export const JWT_EXPIRE_TIME = parseInt(process.env.JWT_EXPIRE_TIME);
export const REFRESH_EXPIRE_TIME = parseInt(process.env.REFRESH_EXPIRE_TIME);
