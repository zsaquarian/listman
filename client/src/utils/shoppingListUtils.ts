import { Storage } from '@capacitor/storage';
import { LIST_SCHEMA_VERSION_NO } from './constants';

export interface ShoppingListItem {
  item: string;
  done: boolean;
}

export interface ShoppingList {
  version: string;
  name: string;
  isShared: boolean;
  items: ShoppingListItem[];
}

export type Lists = {
  key: string;
  name: string;
  isShared: boolean;
  isExternal: boolean;
}[];

export const storeList = async (key: string, list: ShoppingList): Promise<void> => {
  const value = JSON.stringify(list);
  await Storage.set({ key, value });
};

export const loadList = async (key: string): Promise<ShoppingList> => {
  const { value: val } = await Storage.get({ key });
  console.log(val);
  if (val !== null && val) {
    return JSON.parse(val) as ShoppingList;
  } else {
    throw new Error('List does not exist');
  }
};

export const createList = (vals: ShoppingListItem[] = []): ShoppingList => {
  return { version: LIST_SCHEMA_VERSION_NO, items: vals, name: '', isShared: false };
};

export const addItem = async (list: ShoppingList, newVal: string): Promise<ShoppingList> => {
  if (!list.version) {
    list = createList();
  }

  list.items.push({ item: newVal, done: false });

  return list;
};

export const getLists = async (): Promise<Lists> => {
  const { keys } = await Storage.keys();
  const names: Lists = [];

  for (const key of keys) {
    const { name, isShared } = await loadList(key);

    names.push({ name, key, isShared, isExternal: false });
  }

  return names;
};

export const listToString = (list: ShoppingList) => {
  let ret = `${list.name} list is being shared with you from Listman app
 `;
  list.items.forEach((val) => {
    ret += `${val.done ? '[x]' : '[ ]'} ${val.item}
`;
  });
  return ret;
};
