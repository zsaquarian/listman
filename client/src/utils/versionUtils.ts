import { DateTime } from 'luxon';
import { LIST_SCHEMA_VERSION_NO } from './constants';
import type { GenericList } from './listUtils';

export interface ListVersion {
  major: number;
  minor: number;
}

export const getVersion = (version: string): ListVersion => {
  const majorAndMinor = version.split('.');
  const major = parseInt(majorAndMinor[0]);
  const minor = parseInt(majorAndMinor[1]);

  return { major, minor };
};

export const CURRENT_VERSION_NO = getVersion(LIST_SCHEMA_VERSION_NO);

export const getListVersion = (list: GenericList): ListVersion => {
  return getVersion(list.version);
};

export const migrateList = (list: GenericList): GenericList => {
  const version = getListVersion(list);

  if (version.major == CURRENT_VERSION_NO.major) {
    list.version = LIST_SCHEMA_VERSION_NO; // update the minor version number, if needed

    if (!list.modified) {
      list.modified = DateTime.now();
    }

    return list;
  } else {
    // TODO: put in migration code for new major versions
  }

  return list;
};
