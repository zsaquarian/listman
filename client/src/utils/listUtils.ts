import { Storage } from '@capacitor/storage';
import { IGNORED_KEYS, LIST_SCHEMA_VERSION_NO } from './constants';
import { DateTime } from 'luxon';
import { migrateList } from './versionUtils';

export interface GenericListItem {
  item: string;
  done: boolean;
}

export interface GenericList {
  version: string;
  name: string;
  isShared: boolean;
  items: GenericListItem[];
  modified: string;
}

export type Lists = {
  key: string;
  name: string;
  isShared: boolean;
  isExternal: boolean;
}[];

export const storeList = async (key: string, list: GenericList): Promise<void> => {
  list.modified = DateTime.now().toISO();

  const value = JSON.stringify(list);
  await Storage.set({ key, value });
};

export const loadList = async (key: string): Promise<GenericList> => {
  const { value: val } = await Storage.get({ key });
  if (val !== null && val) {
    let list = JSON.parse(val) as GenericList;

    list = migrateList(list);

    return list;
  } else {
    throw new Error('List does not exist');
  }
};

export const loadOrCreateList = async (key: string): Promise<GenericList> => {
  try {
    const loadedList = await loadList(key);
    return loadedList;
  } catch (e) {
    return createList();
  }
};

export const createList = (vals: GenericListItem[] = []): GenericList => {
  return { version: LIST_SCHEMA_VERSION_NO, items: vals, name: '', isShared: false, modified: DateTime.now() };
};

export const addItem = async (list: GenericList, newVal: string): Promise<GenericList> => {
  if (!list.version) {
    list = createList();
  }

  list.items.push({ item: newVal, done: false });

  return list;
};

export const getLists = async (): Promise<Lists> => {
  const { keys } = await Storage.keys();
  const names: Lists = [];

  for (const key of keys) {
    if (IGNORED_KEYS.includes(key)) continue;

    const { name, isShared } = await loadList(key);

    names.push({ name, key, isShared, isExternal: false });
  }

  return names;
};

export const listToString = (list: GenericList): string => {
  let ret = `${list.name} list is being shared with you from Listman app
 `;
  list.items.forEach((val) => {
    ret += `${val.done ? '[x]' : '[ ]'} ${val.item}
`;
  });
  return ret;
};

export const getFormattedModifiedTime = (list: GenericList): string => {
  const modifiedTime = DateTime.fromISO(list.modified);

  console.log(list.modified);

  const ago = modifiedTime.diffNow();
  // -1 is there as luxon does `now - modified`, but we want `modified - now`
  const minutesAgo = -1 * ago.as('minutes');

  if (minutesAgo < 1) {
    return 'a few seconds ago';
  }

  if (minutesAgo < 60) {
    return `${minutesAgo.toFixed(0)} minutes ago`;
  }

  const hoursAgo = -1 * ago.as('hours');

  if (hoursAgo < 24) {
    return `${hoursAgo.toFixed(0)} hours ago`;
  }

  const daysAgo = -1 * ago.as('days');

  if (daysAgo < 7) {
    return `${daysAgo.toFixed(0)} days ago`;
  }

  return ago.toFormat('dd, MM yyyy');
};

export const isDifferent = (list1: GenericList, list2: GenericList): boolean => {
  const list1Copy = { ...list1, modified: undefined };
  const list2Copy = { ...list2, modified: undefined };

  console.log(JSON.stringify(list1Copy) !== JSON.stringify(list2Copy));

  return JSON.stringify(list1Copy) !== JSON.stringify(list2Copy);
};
