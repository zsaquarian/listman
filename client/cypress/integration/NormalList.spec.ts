import type { GenericList } from '../../src/utils/listUtils';
import { LIST_SCHEMA_VERSION_NO } from '../../src/utils/constants';
import { DateTime } from 'luxon';
import { Settings } from 'luxon';

// it's always Apr 1, 2022
Settings.now = () => new Date(2022, 3, 1).valueOf();

describe('Normal List', () => {
  const key = 'test-key';
  const capKey = 'CapacitorStorage.' + key;
  const time = DateTime.now().toISO();
  const testList = {
    name: 'test list',
    version: LIST_SCHEMA_VERSION_NO,
    isShared: false,
    modified: time,
    items: [{ item: 'test item', done: false }],
  } as GenericList;

  before(() => {
    cy.visit('/');
    window.localStorage.setItem(capKey, JSON.stringify(testList));
  });

  it('should load the key', () => {
    cy.visit(`/list/${key}`);
    expect(cy.contains('test item'));
  });

  describe('should save changes', () => {
    it('when adding items', () => {
      cy.get('.input').type('new item');
      cy.get('.absolute > .ml-auto').click();
      cy.getLocalStorage(capKey)
        .then((val) => JSON.parse(val))
        .toMatchSnapshot();
    });

    it('when changing the title', () => {
      cy.get('.text-4xl').clear().type('new title');
      cy.getLocalStorage(capKey)
        .then((val) => JSON.parse(val))
        .toMatchSnapshot();
    });

    it('when toggling items', () => {
      cy.get('div.w-full > :nth-child(2) > .mr-2').click();
      cy.getLocalStorage(capKey)
        .then((val) => JSON.parse(val))
        .toMatchSnapshot();
    });
  });
});
