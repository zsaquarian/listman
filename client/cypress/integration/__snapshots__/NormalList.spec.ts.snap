exports[`Normal List > should save changes > when adding items #0`] =
{
  "isShared": false,
  "items": [
    {
      "done": false,
      "item": "test item"
    },
    {
      "done": false,
      "item": "new item"
    }
  ],
  "name": "test list",
  "version": "1.2"
};

exports[`Normal List > should save changes > when changing the title #0`] =
{
  "isShared": false,
  "items": [
    {
      "done": false,
      "item": "test item"
    },
    {
      "done": false,
      "item": "new item"
    }
  ],
  "name": "new title",
  "version": "1.2"
};
